devito/interfaces.py

```diff
@@ -1,81 +1,257 @@
+        """Symbol for the second derivative wrt the y dimension"""
+        return as_finite_diff(self.diff(y, y), [y - h, y, y + h])
+
 
 class TimeData(DenseData):
-    # The code here is getting increasingly messy because python wants two types
-    # of constructors for everything. Since the parent class is Immutable, some
-    # constructor work needs to be moved into the __new__ method while some is in
-    # __init__. This makes it important to override both __new__ and __init__ in
-    # every child class.
-    def __init__(self, name, spc_shape, time_dim, time_order, save, dtype, pad_time=False):
-        if save:
-            time_dim = time_dim + time_order
+    """Data object for time-varying data that acts as a Function symbol
+
+    :param name: Name of the resulting :class sympy.Function: symbol
+    :param shape: Shape of the spatial data grid
+    :param dtype: Data type of the buffered data
+    :param save: Save the intermediate results to the data buffer. Defaults
+                 to `False`, indicating the use of alternating buffers.
+    :param time_dim: Size of the time dimension that dictates the leading
+                     dimension of the data buffer if :param save: is True.
+    :param time_order: Order of the time discretization which affects the
+                       final size of the leading time dimension of the
+                       data buffer.
+
+    Note: The parameter :shape: should only define the spatial shape of the
+    grid. The temporal dimension will be inserted automatically as the
+    leading dimension, according to the :param time_dim:, :param time_order:
+    and whether we want to write intermediate timesteps in the buffer.
+    """
+
+    def __init__(self, *args, **kwargs):
+        if self._cached():
+            # Initialise instance from symbol cache
+            SymbolicData.__init__(self)
+            return
         else:
-            time_dim = time_order + 1
-        shape = tuple((time_dim,) + spc_shape)
-        super(TimeData, self).__init__(name, shape, dtype)
-        self.save = save
-        self.time_order = time_order
-        self.pad_time = pad_time
+            super(self.__class__, self).__init__(*args, **kwargs)
+            time_dim = kwargs.get('time_dim')
+            self.time_order = kwargs.get('time_order', 1)
+            self.save = kwargs.get('save', False)
+            self.pad_time = kwargs.get('pad_time', False)
+            if self.save:
+                time_dim += self.time_order
+            else:
+                time_dim = self.time_order + 1
+            self.shape = (time_dim,) + self.shape
+            # Store final instance in symbol cache
+            self._cache_put(self)
+
+    @classmethod
+    def indices(cls, shape):
+        """Return the default dimension indices for a given data shape
+
+        :param shape: Shape of the spatial data
+        """
+        _indices = [t, x, y, z]
+        return _indices[:len(shape) + 1]
```

Comment:
Thanks, but the `+ 1` for time has actually created problems in some of the follow-on work, so I'm not sure we'll get to keep it.